commands/command_track.go

```diff
@@ -226,6 +230,16 @@ ArgsLoop:
 	err = lockClient.FixFileWriteFlagsInDir(relpath, readOnlyPatterns, writeablePatterns)
 	if err != nil {
 		LoggedError(err, tr.Tr.Get("Error changing lockable file permissions: %s", err))
+		sawError = true
+	}
+
+	if sawError {
+		os.Exit(2)
+	} else if !modified && len(changedAttribLines) > 0 {
+		// If we didn't modify things, but that's because the patterns
+		// were already supported, don't return an error, since what the
+		// user wanted has already been done.
+		os.Exit(1)
```

Comment:
Is it the case that we would get here if the requested patterns are already supported?  IIUC, the loop which [checks](https://github.com/git-lfs/git-lfs/blob/ab3f2288bdb3cface5a4a0af2daa9c302214ab6f/commands/command_track.go#L96-L97) for that case returns before [adding](https://github.com/git-lfs/git-lfs/blob/ab3f2288bdb3cface5a4a0af2daa9c302214ab6f/commands/command_track.go#L107) anything to `changedAttribLines`.  We would get here, though, if a pattern is [reported](https://github.com/git-lfs/git-lfs/blob/ab3f2288bdb3cface5a4a0af2daa9c302214ab6f/commands/command_track.go#L200) as matching a forbidden file, I think.

Should the comment reflect this, and that we are returning an error (i.e., a non-zero exit code)?  Or am I misunderstanding the context?

Reply:
The comment refers to the condition above it.  Would you prefer that I place it above the condition instead?  Unfortunately, Go effectively requires cuddling the brace and the else if, so I can't split them here.

Reply:
I've moved this out into its own separate `if` block to make the comment cleaner.

And yes, without the `len(changedAttribLines) > 0` condition, one of our other tests fails.

Reply:
Thanks!  I understood that the comment applied to the condition of the enclosing `if` statement block, but it's definitely even clearer now.

What I'm perhaps fuzzy about is how exiting with a `1` counts as "not returning an error".  Is the intent of this comment to state "we don't return a Go error to the caller" and not "we don't exit with a non-zero exit code, which indicates a general error, from the command itself", which is how I've been interpreting it?

And, separately, I still wonder if it is actually possible to get into this conditional if, as the comment says, "we didn't modify things, but that's because the patterns were already supported".

I see that, for instance, that without this [conditional](https://github.com/git-lfs/git-lfs/blob/8e14a7e89a6ba03f6139e88b3834892fc944e184/commands/command_track.go#L239-L244), the `blocklisted` tests fail when `git lfs track .gitattributes` is [run](https://github.com/git-lfs/git-lfs/blob/8e14a7e89a6ba03f6139e88b3834892fc944e184/t/t-track.sh#L380) or `git lfs track .git*` is [run](https://github.com/git-lfs/git-lfs/blob/8e14a7e89a6ba03f6139e88b3834892fc944e184/t/t-track.sh#L399) and those commands don't return a non-zero (i.e., error) exit code.  That makes sense to me.  When the conditional exists, the tests pass.

But IIUC the situation caught by this conditional block is not one where nothing is being changed because "the patterns were already supported", as the comment states currently.  Rather, nothing is being changed because the patterns are disallowed -- the code would change something, but it's noticing that we're not supposed to change `.git*` or `.gitattributes`, and causing the command to exit with a non-zero error code (which is good!)

It seems to me that the case where we don't change any patterns because the patterns were already supported, like in the example below, is not passing through this conditional.  Rather, the condition is [caught](https://github.com/git-lfs/git-lfs/blob/8e14a7e89a6ba03f6139e88b3834892fc944e184/commands/command_track.go#L96-L97) in a prior loop over the input command-line arguments, and the `continue` in that conditional block means nothing is [added](https://github.com/git-lfs/git-lfs/blob/8e14a7e89a6ba03f6139e88b3834892fc944e184/commands/command_track.go#L107) to `changedAttribLines` since that occurs later in the loop.  Thus when we get to the conditional under discussion here (and assuming there aren't any valid or blocklisted patterns to consider among the input arguments),  `modified` will be false, but `changedAttribLines` will also just be empty, and this conditional doesn't apply.
```
$ git lfs track "*.bin"
Tracking "*.bin"
$ git lfs track "*.bin"
"*.bin" already supported
```

Does that all make sense?

Reply:
If we had just `if !modified` for this block, then we'd return 1 if we had no modified lines because they were all disallowed.  However, that condition would also be triggered if we had no modified lines because all of our patterns were already supported.  The way we know that has occurred is if `changedAttribLines` is empty, since, as you mention, the previous loop will have skipped adding them.

So, with this comment, we're describing part of the behaviour of the implicit and empty else branch.  It's true that exiting 1 is, indeed, returning an error.  However, if we don't take the if-branch, then we return 0 implicitly, which is the non-error case, and what the comment describes.  I agree the comment is confusing if we read it as applying to the if branch.

Hopefully that makes things clearer here.  If you think that the comment should be reworded to refer to the if-branch instead in a clearer way, I can do that.  Maybe something like this would be better:

> If we didn't modify anything successfully, only return an error if we have at least one new or changed pattern that isn't alredy supported.  If all of the patterns were already supported and we didn't modify anything for that reason, we don't consider that an error case and skip the conditional below, implicitly returning 0, because what the user wanted has already been done.

Of course I'm happy to hear other suggestions about how to make this less confusing if you have any.

Reply:
Ah, I see -- the comment is intended to apply to the case where we _don't_ enter the conditional block that exits with `os.Exit(1)`!  That makes it make much more sense to me.

I suppose I might just suggest adding a sentence following the comment, because (and maybe this is just me) I read a comment above or within a conditional block as applying to that block.  Let me draft something in a suggestion real quick.